ecl/hql/hqlattr.cpp

```diff
@@ -2046,37 +2046,56 @@ static unsigned estimateRowSize(IHqlExpression * record)
 }
 
 
-bool reducesRowSize(IHqlExpression * expr)
+static bool increasesRowSize(IHqlExpression * newRecord, IHqlExpression * oldRecord)
 {
-    //More: This should be improved...., but slightly tricky without doing lots more processing.
-    IHqlExpression * transform = queryNewColumnProvider(expr);
-    IHqlExpression * prevRecord = expr->queryChild(0)->queryRecord();
-    unsigned newRowSize = estimateRowSize(transform->queryRecord());
-    unsigned prevRowSize = estimateRowSize(prevRecord);
-    if ((newRowSize != UNKNOWN_LENGTH) && (prevRowSize != UNKNOWN_LENGTH))
-        return newRowSize < prevRowSize;
+    unsigned newRowSize = estimateRowSize(newRecord);
+    unsigned oldRowSize = estimateRowSize(oldRecord);
+    if ((newRowSize != UNKNOWN_LENGTH) && (oldRowSize != UNKNOWN_LENGTH))
+        return newRowSize > oldRowSize;
 
-    IHqlExpression * record = expr->queryRecord();
-    if (getFlatFieldCount(record) < getFlatFieldCount(prevRecord))
+    if (isFixedSizeRecord(newRecord) && newRowSize < getMinRecordSize(oldRecord))
+        return false;
+    if (isFixedSizeRecord(oldRecord) && oldRowSize < getMinRecordSize(newRecord))
         return true;
+
+    HqlRecordStats oldStats;
+    HqlRecordStats newStats;
+    gatherRecordStats(oldStats, oldRecord);
+    gatherRecordStats(newStats, newRecord);
+    if (newStats.fields > oldStats.fields)
+        return true;
+    if (newStats.datasetFields > oldStats.datasetFields)
+        return true;
+    if (newStats.dictionaryFields > oldStats.dictionaryFields)
+        return true;
+
     return false;
 }
 
-bool increasesRowSize(IHqlExpression * expr)
+bool reducesRowSize(IHqlExpression * expr)
```

Comment:
The set of tests used to check if the row size is likely to be similiar is a different set to increasesRowSize.  Should the same set of tests be used?  If not may be worth documenting why increasesRowSize isn't inverse to decreasesRowSize